README.md
# Let Google Firebase create and authenticate users to your Laravel API (using Laravel Passport)

[![Latest Version on Packagist](https://img.shields.io/packagist/v/square1/laravel-passport-firebase-auth.svg?style=flat-square)](https://packagist.org/packages/square1/laravel-passport-firebase-auth)
[![MIT Licensed](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Tests](https://github.com/square1-io/laravel-passport-firebase-auth/workflows/Tests/badge.svg?style=flat-square)](https://github.com/square1-io/laravel-passport-firebase-auth/actions?query=workflow%3ATests+branch%3Amaster)


Create and authenticate users with Firebase Auth providers (Google, Facebook, Apple, email, etc), and let Laravel Passport know and handle your backend secure endpoints!

This is an opinionated way to create Laravel Passport tokens from a Firebase valid token.

## Installation

You can install the package via composer:

```bash
composer require square1/laravel-passport-firebase-auth
```

You will need a `firebase_uid` column on your users table. You can publish and run the migrations and customize that column with:

```bash
php artisan vendor:publish --provider="Square1\LaravelPassportFirebaseAuth\LaravelPassportFirebaseAuthServiceProvider" --tag="migrations"

# after column customization run:
php artisan migrate
```

Publish the config file with:
```bash
php artisan vendor:publish --provider="Square1\LaravelPassportFirebaseAuth\LaravelPassportFirebaseAuthServiceProvider" --tag="config"
```

This is the contents of the published config file:

```php
return [
    /**
     * We will load all the package routes under your api prefix for consistency
     */
    'api_prefix' => env('API_PREFIX', 'api/v1'),

    /**
     * Declare the amount of minutes in the future. Laravel Passport will create a token that endures
     * that long. See: https://laravel.com/docs/7.x/passport#token-lifetimes
     */
    'token_expiration_in_minutes' => 60 * 24 * 7, // Default 1 week expiration

    /**
     * Please update the column names in the array (the values) to match your users database columns.
     * Remove all the key pairs you don't need to store, but leave 'uid' and 'email', we need them!
     * For the 'uid' create a nullable string (we'll do this if use this package migrations)
     * Of course, make all columns fillable in your User model
     * or we can't save them if those columns are protected.
     */
    'map_user_columns' => [
        'uid' => 'firebase_uid', // REQUIRED
        'email' => 'email', // REQUIRED
        'displayName' => 'name',
        'emailVerified' => 'email_verified_at',
        'phoneNumber' => 'phone',
        'photoURL' => 'avatar',
        'provider' => 'provider' // e.g facebook, google, password
    ],

    /**
     * If you need some mandatory columns in order to store your laravel user model,
     * you can indicate them in here along with the validation rules you need.
     * Of course, you will need to post this keys as indicated in the readme.
     * And of course, make all columns fillable in your User model,
     * or we can't save them if those columns are protected.
     */
    'extra_user_columns' => [
        // 'username' => 'required|unique:users|max:255',
        // 'birthday' => 'require|date|before:today|date_format:Y-m-d'
    ],

    /**
     * Please indicate if anonymous firebase users capability is alowed in your application.
     * If so, we will need to olso create an "anonymous" user in your database in order
     * to let Laravel Passport issue a token for that particular user.
     */
    'allow_anonymous_users' => false,

    /**
     * Indicate the mandatory fields for anonymous user creation in your laravel database.
     * For the email, we'll concatenate (prefix) with the UID from firebase.
     */
    'anonymous_columns' => [
        'email' => '@anonymous.com', // This will end up being firebasetoken@anonymous.com
        'name' => 'Anonymous',
        'anonymous' => true,
        // 'avatar' => 'sample_anonymous_avatar.png'
    ]
];
```

### Configure Laravel Passport

This package has Laravel Passport as a dependency, if you did not already, please [configure Laravel Passport](https://laravel.com/docs/7.x/passport).

### Configure Firebase

Create a Firebase project in the console [https://console.firebase.google.com/](https://console.firebase.google.com/).

If you did not already please generated your Service Account auth file, do it from this url: [https://console.firebase.google.com/project/_/settings/serviceaccounts/adminsdk](https://console.firebase.google.com/project/_/settings/serviceaccounts/adminsdk). You will be asked to select the Firebase Project.
After that, the Firebase Admin SDK screen will ask you to pick a language, just leave `Node.js` selected and click `Generate new private key`.

Once you have downloaded the Service Account JSON file in your project (**ATENTION! please git ignore this file** as it has sensible credentials), indicate the path to your file in `.env` like this:

```
FIREBASE_CREDENTIALS=storage/firebase_credentials.json
```

#### Configure auth providers

In your firebase project create and configure all providers you want to use: [https://firebase.google.com/docs/auth](https://firebase.google.com/docs/auth)

## Usage

This package will expose 2 endpoints under your api prefix (configurable):

1) **POST**: `api/v1/firebase/user/create` 

In your mobile app or front end, you will allow your users to create an account using the [Firebase client SDK of your choice](https://firebase.google.com/docs/firestore/client/libraries).

Then you will call this endpoint with a valid firebase token using the `firebase_token` key in the payload posted.

This endpoint will reach firebase database, find and validate the user just created in your front end / mobile app, and it will create a user record in your laravel database saving the `firebase_uid` in users table you populated previously in the installation step.

Optionaly, you can perform 2 extra user configuration steps here:

**1 - a) Conect extra user data from the firebase users payload:**
    
In your config/laravel-passport-firebase-auth.php indicate the keys you want to match against your laravel users table using the "map_user_columns" key in the array.

**1 - b) Pass any other custom data you need for the user creation proces in your laravel database:**

An example will be if user creation require some mandatory columns (e.g. user_plan, username, role, etc.). For this you will use the instructions on the "extra_user_columns" key in the config array.

For security reasons, we'll validate this data, and we'll ignore any other values not declared in this "extra_user_columns" array.

----
Example payload posted to `api/v1/firebase/user/create`:

```json
{
    "firebase_token": "super_long_firebase_token_here",
    "username": "tonystark",
    "plan": "platinum",
    "role": "superadmin"
}
```

if in your `config/laravel-passport-firebase-auth.php` file you have the followin configuration:

```php
    'map_user_columns' => [
        'uid' => 'firebase_uid',
        'email' => 'email',
        'displayName' => 'full_name',
        'photoURL' => 'avatar',
    ],
    'extra_user_columns' => [
        'username' => 'required|unique:users|max:255',
        'plan' => 'required|in:silver,gold,platinum'
    ]
```

The result will be that, the newly created firebase user will be stored in your database with the uid, email, displayName as the full_name column, photoURL as the avatar column, and the rest of the firebase metadata will be discarted.

Also the username and plan will be stored, but the `role` manipulation attempt will be ignored.

You will receive a `success` status from the endpoint, along with the backend user ID and valid Laravel Passport access token.

```json
{
    "status": "success",
    "data": {
        "user_id": 1,
        "access_token": "valid_laravel_passport_token",
        "token_type": "Bearer",
        "expires_at": "2020-09-14T23:16:35.000000Z"
    }
}
```

2) **POST**: `api/v1/firebase/user/login` 

You will need to call this endpoint with only a valid firebase token using the key `firebase_token` in the payload posted.

In case we find the user in the laravel database, the result will contain a `success` status along with the passport token to use in furter requests.

```json
{
    "status": "success",
    "data": {
        "user_id": 1,
        "access_token": "valid_laravel_passport_token",
        "token_type": "Bearer",
        "expires_at": "2020-09-14T23:17:02.000000Z"
    }
}
```

## Testing

``` bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Security

If you discover any security related issues, please email emiliano@square1.io instead of using the issue tracker.

## Credits

- [Emiliano Tisato](https://github.com/emilianotisato)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `square1/laravel-passport-firebase-auth` will be documented in this file

## 2.0.0 - 2020-09-09

- initial release supporting both laravel 7 and 8


tests/User.php
<?php

namespace Square1\LaravelPassportFirebaseAuth\Tests;

use Laravel\Passport\HasApiTokens;
use Illuminate\Foundation\Auth\User as Authenticatable;

class User extends Authenticatable
{
    use HasApiTokens;

    public $timestamps = false;

    protected $table = 'users';

    /**
     * The attributes that are mass assignable.
     *
     * @var array
     */
    protected $fillable = [
        'email', 'firebase_uid', 'name', 'avatar', 'role', 'username',
    ];
}


tests/TestCase.php
<?php

namespace Square1\LaravelPassportFirebaseAuth\Tests;

use Laravel\Passport\Passport;
use Illuminate\Database\Schema\Blueprint;
use Illuminate\Contracts\Config\Repository;
use Laravel\Passport\PassportServiceProvider;
use Orchestra\Testbench\TestCase as Orchestra;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Kreait\Laravel\Firebase\ServiceProvider as FirebaseServiceProvider;
use Square1\LaravelPassportFirebaseAuth\LaravelPassportFirebaseAuthFacade;
use Square1\LaravelPassportFirebaseAuth\LaravelPassportFirebaseAuthServiceProvider;

class TestCase extends Orchestra
{
    use RefreshDatabase;

    public function setUp(): void
    {
        parent::setUp();

        $this->withFactories(__DIR__.'/database/factories');

        $this->setUpDatabase($this->app);

        Passport::routes();

        $this->artisan('passport:install');
    }

    public function getEnvironmentSetUp($app)
    {
        $config = $app->make(Repository::class);
        $config->set('auth.defaults.provider', 'users');
        $config->set('auth.guards.api', ['driver' => 'passport', 'provider' => 'users']);

        $app['config']->set('auth.providers.users.model', 'Square1\LaravelPassportFirebaseAuth\Tests\User');

        $app['config']->set('database.default', 'sqlite');

        $app['config']->set('passport.storage.database.connection', 'sqlite');

        $app['config']->set('app.key', 'base64:fakekey/avhnnoiIltExLrEfZvvZx7h1Hb29Pgel2ec=');
        $app['config']->set('passport.public_key', $this->getFakePublicOauth());
        $app['config']->set('passport.private_key', $this->getFakePrivateOauth());

        $app['config']->set('database.connections.sqlite', [
            'driver' => 'sqlite',
            'database' => ':memory:',
            'prefix' => '',
        ]);
    }

    /**
     * @param \Illuminate\Foundation\Application $app
     *
     * @return array
     */
    protected function getPackageProviders($app)
    {
        return [
            PassportServiceProvider::class,
            FirebaseServiceProvider::class,
            LaravelPassportFirebaseAuthServiceProvider::class,
        ];
    }

    /**
     * @param \Illuminate\Foundation\Application $app
     *
     * @return array
     */
    protected function getPackageAliases($app)
    {
        return [
            'LaravelPassportFirebaseAuth' => LaravelPassportFirebaseAuthFacade::class,
        ];
    }

    /**
     * Set up the database.
     *
     * @param \Illuminate\Foundation\Application $app
     */
    protected function setUpDatabase($app)
    {
        $this->artisan('migrate:fresh');

        // Create a light weight version of users table
        $app['db']->connection()->getSchemaBuilder()->create('users', function (Blueprint $table) {
            $table->increments('id');
            $table->string('email');
            $table->string('name')->nullable();
            $table->string('avatar')->nullable();
            $table->string('username')->nullable();
            $table->string('role')->nullable();
        });

        include_once __DIR__.'/../database/migrations/create_laravel_passport_firebase_auth_table.php.stub';
        (new \CreateLaravelPassportFirebaseAuthTable())->up();
    }

    public function getFakePublicOauth()
    {
        return "-----BEGIN PUBLIC KEY-----\nMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAs7O3V2CP9NOsB9iNl4H1\nbQmZpUHjlV2O6HNGM+3yhy1a9BB16gOpqcegykpVqqaxLDqf1MLX2LSVUAdlREZK\ndNUKcM9t8V9On0mRpgL3GDFCFzO5TMTXxioUfq6HhF0fukYQ/Gx9f827eAuCJ1An\nuF9kYLKuYAxt6DpqTzuEvOcjoNbzmIl1dvnQFCODDZJfjSZfOwuMn9WeQATyL36w\nd487fhaeeDRUka5UfHoS1DUdnlfVrFyZw3NiwGI4ykuvQ8sbE0PYiVi0WsPfy00Y\neCg4olcsTu8+gwHZ3MbD5CeGXBdAEtZSLTFxmz/7OiZIrhg4JeLyJNNIxmQ8DVJY\nAJTYyulAUea0e0IuX6QJNbv4WHc08hNDj+XV5r9t19qyTL/Oa5PIWxcDhW1RHDjM\n97fcesKe/xibA+aiWUUjl4Mv/QevWTM7yJke5cVQyuuyPKUdXrl+EQ21J4+4S1B4\nS/ouGlUJhd14uVsjJwojPd6iELjLKHRQ2lwe/AP+B5UKmtQyJ+uMZLkFT0FbeSXw\ntbN9tN+RHdTqPbgCnZTzVeqMGgh0bukRzRpMJIX/ik2rsDy/VY2LTdGLoBFsOpqN\nelQcd/N56pD4ldvQUAb814nTO+CcxRsY5a4xWqrONO4xFJt906Ms8Bpo+AVITfBQ\njFNe2BGK57CPm9umylg8cwMCAwEAAQ==\n-----END PUBLIC KEY-----";
    }

    public function getFakePrivateOauth()
    {
        return "-----BEGIN RSA PRIVATE KEY-----\nMIIJKAIBAAKCAgEAs7O3V2CP9NOsB9iNl4H1bQmZpUHjlV2O6HNGM+3yhy1a9BB1\n6gOpqcegykpVqqaxLDqf1MLX2LSVUAdlREZKdNUKcM9t8V9On0mRpgL3GDFCFzO5\nTMTXxioUfq6HhF0fukYQ/Gx9f827eAuCJ1AnuF9kYLKuYAxt6DpqTzuEvOcjoNbz\nmIl1dvnQFCODDZJfjSZfOwuMn9WeQATyL36wd487fhaeeDRUka5UfHoS1DUdnlfV\nrFyZw3NiwGI4ykuvQ8sbE0PYiVi0WsPfy00YeCg4olcsTu8+gwHZ3MbD5CeGXBdA\nEtZSLTFxmz/7OiZIrhg4JeLyJNNIxmQ8DVJYAJTYyulAUea0e0IuX6QJNbv4WHc0\n8hNDj+XV5r9t19qyTL/Oa5PIWxcDhW1RHDjM97fcesKe/xibA+aiWUUjl4Mv/Qev\nWTM7yJke5cVQyuuyPKUdXrl+EQ21J4+4S1B4S/ouGlUJhd14uVsjJwojPd6iELjL\nKHRQ2lwe/AP+B5UKmtQyJ+uMZLkFT0FbeSXwtbN9tN+RHdTqPbgCnZTzVeqMGgh0\nbukRzRpMJIX/ik2rsDy/VY2LTdGLoBFsOpqNelQcd/N56pD4ldvQUAb814nTO+Cc\nxRsY5a4xWqrONO4xFJt906Ms8Bpo+AVITfBQjFNe2BGK57CPm9umylg8cwMCAwEA\nAQKCAgAkTdJkXKW4mGrQyvcP/LlQZfgcYstPia8tVtx/8TpmBMuzMwAfXs4P9ryN\nIadc6oAwp0dS/GoO5aykllnnCSxRnhiV4dIcSVzg4UQDfeXdhVYMye5NjBbreeTa\nEvhdzVJzl2QnGuPfxfhxsCGDP7ZGkT7+KhsAXIJ3wBVqHGQcpbWU8NgVoO+SMbXP\n27zGKSQUqAPlMdgfElD+LgGfhsCv5sfmGTu6nRfYtpdq7l6PcIujSatpPuvxiIYS\nx8UhWVj3ITk/Ex3T4Y/OJnQ35kNjQvzuDx5sc/j1DpKs/5rjpPiH8kqHg7TmXG7P\nJoXioYldYpc1UDwLJIQ3UFixeEk+viahCACsy/3J/YnnF2diKNT/VKd8UTPIu8qJ\n9E+3Z40KykumL3v5Sk0O38P8r6RaG1xjV9sLk/jHT8UskVBDe1tYQ3LQqTESvajH\neOBKeBYRYEH6/SJTq/IW2Lv6ge0MzNCX820SHld13eNsgU+ql5HuEFMrtOY7Hg1t\nN43pS3JecTIH1bWvW+zSF03+4Oy8HfiK5Wwz+fbpKvM1ZgUBMJRoYSw//geUYNzV\n0HbdDo7+MR8cNplir/42ZxtM73ot72aOxQImCfw1F38LCtieyz+GG2c1Juk/nSM/\n2bTuqDfUgljLN1H29ipU+Hx8Q5aa3OaL6royq9Y79uuX6n+sQQKCAQEA2g+FZbRX\nvGKa2yvOjmCENyKazEd/aqVy+7zppkhdzjqEiw36ntO+5fD9mKPJmnZRh41bxSwR\nO7hYnmbR8Vm9aFc559KG2p8kI+QGcZGGx5eRJhAtsprJEax+1x5v6EFJHUhrrW42\nmYndwxKOEsx/pkY6/bPfXq3RPPnHJb3zGf/bgly+pJECe1l2GbQRWPDnmXuacV25\nyJ1eHSLBgka3/hyiHHb0YeDXlLG+IYgmiEpM+O2W/+Vd0FmKNrJVJX68lbPvizYe\nfY6xDqbqhkV2Q0nLkA6W9F2TtUH4QIpoe6hglL6giZ+kWK9ZcEeefBiASwHJJQwj\nIbqOcB8fWo5L9wKCAQEA0vexMxQo8afrUg11rjSmr9pg03owwCCgtl7LUAbHGErg\n99Y+AXFslS8fymmj6Vbsusd9Q7TB87G5z813HWbDJE4Yr/eoMehcZneQQJIwVZeg\nFQdnzT44Fei9jHY3o3eJ3WlhPFB9mq323pNCGPqhnlynb+QphyL1PUxHLdn5E1Vs\nL6O89zavFk0Q6mXa59gb2s67VwQH5H2xUZ/4C/n5O63xFdHV5sshmRHXNVoz3oPx\n4VarG4nqQTeHwLTeScN18L9//EdLw43MSjhaBQ3Q4bdacqM01sqpjdbZHZq5Kspa\nHRLGgqLUfmm+FTTgVRJ9MVPC8y7Y3ABeEJXEPboWVQKCAQEAowfTrjRlHB2EtbRM\n3Dng3+/pWC1kX/GOxBN7hKy6JYOusOAkrjaVlQjWMRbTlb48OmI/aG0H/WRYLSWm\ndRGpAKemTWIjHLS6qnLDNomdAQCarqCN6ei7x5D1zBOfiz+0UEZi6ulpvOVMfZoT\nyo5GKaR8Wk3vBLRjqXj9oYQgiyG1lgONLTCVcG90UsdD6QFDxoaY84Ulb43oXVQs\n5R/GVCBIO4vcLomR5EP3aM6IMIGMhtfreyqbNziak0ZFTqEwkaRTxfsSVMEoM1Is\nKXMpdiloHi4qQkzsMCpAVtCkST+b5dmX4Q0QLJX9AmspXJJc6LdPEXm/kmoOMFm0\ncVnOBQKCAQBsp/nyDt4PqKbAugH0WVXImLbp9LMLIULk6unYK8V7M4Wu3/9Livmb\n1IuRGtu7IHQItxpDNuP1+YF0D6Tb1cOH/VkluG4+VijQ9Z+sQh721oMykX/a68LD\nNf36TCDX5odxLAdSozot4o+Vj06pwtxezeXG8UKaQV0B1zmJ2gw48vQnjTOUN+vx\nlnux3gfjBSFDjtpaNM9D1aDwI24D7Rl6rVnQHSUIG3MQfWnUJsM0RczcfGDgbCXk\ntQ8MJ5udbjheaKMocigJbgzb5S4oEDeXKXJCPODIB9VQImnsn3XjHhlPhA4N1oOP\nIDMwhO1No5orP7LWwTgcB2xrKlfKWv+JAoIBADkT2QoeE/7ORuSkul98q3qNk4Ep\npyCgWZsNrrYAEDwpIggE39x7j12XruXQHB7L93gNJUx/XwpEOANl8v3YaU7BjKkP\ndZEXCStmv5cC3Nvr8TRN5w+SxSnrK8MNCjH8Rofvxp5UuJn+ZvCLso2Nslz5t25p\n7y2FKFnbS+l+lf6rZakZq1A2pOPUSQK7BCitnC2CIxUSYbnx4zKs+tV3MnAzG0vA\n47Myx33CivrNME0PCRGHx6bSp8z9sYQUv0gHJKTMuIWu6kD+XYyZxfYjLpwgfioe\n5YCrm2SDfhP8s3UW3A0Q7bQQv86wtiNUM1WqgtArjxpTa0fvgaicPyGAccw=\n-----END RSA PRIVATE KEY-----";
    }
}


tests/ExampleTest.php
<?php

namespace Square1\LaravelPassportFirebaseAuth\Tests;

class ExampleTest extends TestCase
{
    /** @test */
    public function true_is_true()
    {
        $this->assertTrue(true);
    }
}


src/LaravelPassportFirebaseAuth.php
<?php

namespace Square1\LaravelPassportFirebaseAuth;

use Illuminate\Foundation\Auth\User;
use Kreait\Firebase\Auth\UserRecord;
use Laravel\Passport\PersonalAccessTokenResult;

class LaravelPassportFirebaseAuth
{
    /**
     * Find firebase user record
     *
     * @param string $token
     */
    public function getUserFromToken(string $token) : UserRecord
    {
        // Get Kreait\Firebase\Auth instance from the container
        $auth = app('firebase.auth');

        $verifiedIdToken = $auth->verifyIdToken($token);

        // Retrieve the UID (User ID) from the verified Firebase credential's token
        $uid = $verifiedIdToken->getClaim('sub');

        return $auth->getUser($uid);
    }

    /**
     * Create a valid passport token for the given user
     *
     * @param \Illuminate\Foundation\Auth\User $user
     * @return \Laravel\Passport\PersonalAccessTokenResult
     */
    public function createPassportToken(User $user): PersonalAccessTokenResult
    {
        $tokenResult = $user->createToken('Personal Access Token');
        $tokenResult->token->expires_at = now()->addMinutes(config('laravel-passport-firebase-auth.token_expiration_in_minutes'));
        $tokenResult->token->save();

        return $tokenResult;
    }

    /**
     * Check if firebase user is anonymous
     *
     * @param \Kreait\Firebase\Auth\UserRecord $firebaseUser
     * @return bool
     */
    public function isAnonymousUser(UserRecord $firebaseUser) : bool
    {
        // If user has a valid UID but and empty array as provider
        return $firebaseUser->uid && $firebaseUser->providerData == [];
    }
}


src/Http/Controllers/FirebaseAuthController.php
<?php

namespace Square1\LaravelPassportFirebaseAuth\Http\Controllers;

use Illuminate\Support\Str;
use Illuminate\Http\Request;
use LaravelPassportFirebaseAuth;
use Illuminate\Http\JsonResponse;
use Kreait\Firebase\Auth\UserRecord;
use Illuminate\Support\Facades\Validator;
use Firebase\Auth\Token\Exception\InvalidToken;
use Square1\LaravelPassportFirebaseAuth\Exceptions\NoUidColumnDeclaredException;

class FirebaseAuthController
{
    private string $uid_column;

    public function __construct()
    {
        $this->uid_column = config('laravel-passport-firebase-auth.map_user_columns.uid');

        if (! $this->uid_column) {
            throw NoUidColumnDeclaredException::create();
        }
    }

    public function createUserFromFirebase(Request $request) : JsonResponse
    {
        $firebaseUser = LaravelPassportFirebaseAuth::getUserFromToken($request->firebase_token);

        try {
            $is_anonymous = LaravelPassportFirebaseAuth::isAnonymousUser($firebaseUser) && config('laravel-passport-firebase-auth.allow_anonymous_users');
            $data = array_merge(
                $this->buildUserData($firebaseUser, $is_anonymous),
                $request->only(
                    array_keys($extra_user_columns = config('laravel-passport-firebase-auth.extra_user_columns'))
                )
            );

            $authenticable_class = config('auth.providers.users.model');
            /** @psalm-suppress UndefinedClass */
            $usersTable = (new $authenticable_class)->getTable();
            $rules = array_merge($extra_user_columns, [
                $this->uid_column => 'required:unique:'.$usersTable,
                'email' => 'required|unique:'.$usersTable,
            ]);

            $validator = Validator::make($data, $rules);
            if ($validator->fails()) {
                return response()->json(['errors' => $validator->errors()], 422);
            }

            /** @psalm-suppress UndefinedMethod */
            $user = config('auth.providers.users.model')::create($data);
        } catch (\Illuminate\Database\QueryException $e) {
            return response()->json([
                'message' => 'Unauthorized - Can\'t process some database column: '.$e->getMessage(),
            ], 401);
        }

        $tokenResult = LaravelPassportFirebaseAuth::createPassportToken($user);

        return response()->json([
            'status' => 'success',
            'data' => [
                'user_id' => $user->id,
                'access_token' => $tokenResult->accessToken,
                'token_type' => 'Bearer',
                'expires_at' => $tokenResult->token->expires_at,
            ],
        ]);
    }

    public function loginFromFirebase(Request $request) : JsonResponse
    {
        try { // Try to verify the Firebase credential token with Google
            $firebaseUser = LaravelPassportFirebaseAuth::getUserFromToken($request->firebase_token);
        } catch (\InvalidArgumentException $e) { // If the token has the wrong format
            return response()->json([
                'message' => 'Unauthorized - Can\'t parse the token: '.$e->getMessage(),
            ], 401);
        } catch (InvalidToken $e) { // If the token is invalid (expired ...)
            return response()->json([
                'message' => 'Unauthorized - Token is invalide: '.$e->getMessage(),
            ], 401);
        }

        // Retrieve the user model linked with the Firebase UID
        /** @psalm-suppress UndefinedMethod */
        $user = config('auth.providers.users.model')::where($this->uid_column, $firebaseUser->uid)->first();

        if (! $user) {
            return response()->json([
                'message' => 'Unauthorized - User not found for the given firebase credentials.',
            ], 404);
        }

        $tokenResult = LaravelPassportFirebaseAuth::createPassportToken($user);

        return response()->json([
            'status' => 'success',
            'data' => [
                'user_id' => $user->id,
                'access_token' => $tokenResult->accessToken,
                'token_type' => 'Bearer',
                'expires_at' => $tokenResult->token->expires_at,
            ],
        ]);
    }

    private function buildUserData(UserRecord $firebaseUser, $anonymous = false) : array
    {
        $data = [];
        // Add firebase user data
        foreach (config('laravel-passport-firebase-auth.map_user_columns') as $firebaseKey => $column) {
            if (property_exists($firebaseUser, $firebaseKey)) {
                $data[$column] = $firebaseUser->{$firebaseKey};
            }
            if ($firebaseKey == 'provider') {
                if (property_exists($firebaseUser, 'providerData') && $firebaseUser->providerData != null) {
                    $data[$column] = $firebaseUser->providerData[0]->providerId;
                }
            }
            if ($firebaseKey == 'emailVerified') {
                $data[$column] = $firebaseUser->{$firebaseKey} ? now()->format('Y-m-d') : false;
            }
        }

        if ($anonymous) {
            $dataOverride = array_merge(
                config('laravel-passport-firebase-auth.anonymous_columns'),
                [$this->uid_column => $firebaseUser->uid]
            );
            if (key_exists('email', $dataOverride)) {
                $dataOverride['email'] = $firebaseUser->uid.$dataOverride['email'];
            }
        } else {
            $dataOverride = [
                'email' => $firebaseUser->email,
                $this->uid_column => $firebaseUser->uid,
            ];
        }
        $data = array_merge($data, $dataOverride);

        $data['password'] = Str::random(32);

        return $data;
    }
}


src/LaravelPassportFirebaseAuthServiceProvider.php
<?php

namespace Square1\LaravelPassportFirebaseAuth;

use Illuminate\Support\Facades\Route;
use Illuminate\Support\ServiceProvider;
use Square1\LaravelPassportFirebaseAuth\Http\Controllers\FirebaseAuthController;

class LaravelPassportFirebaseAuthServiceProvider extends ServiceProvider
{
    public function boot()
    {
        if ($this->app->runningInConsole()) {
            $this->publishes([
                __DIR__ . '/../config/laravel-passport-firebase-auth.php' => config_path('laravel-passport-firebase-auth.php'),
            ], 'config');

            $this->publishes([
                __DIR__ . '/../resources/views' => base_path('resources/views/vendor/laravel-passport-firebase-auth'),
            ], 'views');

            $migrationFileName = 'create_laravel_passport_firebase_auth_table.php';
            if (! $this->migrationFileExists($migrationFileName)) {
                $this->publishes([
                    __DIR__ . "/../database/migrations/{$migrationFileName}.stub" => database_path('migrations/' . date('Y_m_d_His', time()) . '_' . $migrationFileName),
                ], 'migrations');
            }
        }

        $this->loadViewsFrom(__DIR__ . '/../resources/views', 'laravel-passport-firebase-auth');

        Route::prefix(config('laravel-passport-firebase-auth.api_prefix'))->group(function () {
            Route::post('firebase/user/create', [FirebaseAuthController::class, 'createUserFromFirebase']);
            Route::post('firebase/user/login', [FirebaseAuthController::class, 'loginFromFirebase']);
        });
    }

    public function register()
    {
        $this->mergeConfigFrom(__DIR__ . '/../config/laravel-passport-firebase-auth.php', 'laravel-passport-firebase-auth');

        $this->app->bind(LaravelPassportFirebaseAuth::class, function () {
            return new LaravelPassportFirebaseAuth;
        });

        $this->app->alias(LaravelPassportFirebaseAuth::class, 'laravel-passport-firebase-auth');
    }

    public static function migrationFileExists(string $migrationFileName): bool
    {
        $len = strlen($migrationFileName);
        foreach (glob(database_path("migrations/*.php")) as $filename) {
            if ((substr($filename, -$len) === $migrationFileName)) {
                return true;
            }
        }

        return false;
    }
}


config/laravel-passport-firebase-auth.php
<?php

return [
    /**
     * We will load all the package routes under your api prefix for consistency
     */
    'api_prefix' => env('API_PREFIX', 'api/v1'),

    /**
     * Declare the amount of minutes in the future. Laravel Passport will create a token that endures
     * that long. See: https://laravel.com/docs/7.x/passport#token-lifetimes
     */
    'token_expiration_in_minutes' => 60 * 24 * 7, // Default 1 week expiration

    /**
     * Please update the column names in the array (the values) to match your users database columns.
     * Remove all the key pairs you don't need to store, but leave 'uid' and 'email', we need them!
     * For the 'uid' create a nullable string (we'll do this if use this package migrations)
     * Of course, make all columns fillable in your User model
     * or we can't save them if those columns are protected.
     */
    'map_user_columns' => [
        'uid' => 'firebase_uid', // REQUIRED
        'email' => 'email', // REQUIRED
        'displayName' => 'name',
        'emailVerified' => 'email_verified_at',
        'phoneNumber' => 'phone',
        'photoURL' => 'avatar',
        'provider' => 'provider', // e.g facebook, google, password
    ],

    /**
     * If you need some mandatory columns in order to store your laravel user model,
     * you can indicate them in here along with the validation rules you need.
     * Of course, you will need to post this keys as indicated in the readme.
     * And of course, make all columns fillable in your User model,
     * or we can't save them if those columns are protected.
     */
    'extra_user_columns' => [
        // 'username' => 'required|unique:users|max:255',
        // 'birthday' => 'require|date|before:today|date_format:Y-m-d'
    ],

    /**
     * Please indicate if anonymous firebase users capability is alowed in your application.
     * If so, we will need to olso create an "anonymous" user in your database in order
     * to let Laravel Passport issue a token for that particular user.
     */
    'allow_anonymous_users' => false,

    /**
     * Indicate the mandatory fields for anonymous user creation in your laravel database.
     * For the email, we'll concatenate (prefix) with the UID from firebase.
     */
    'anonymous_columns' => [
        'email' => '@anonymous.com', // This will end up being firebasetoken@anonymous.com
        'name' => 'Anonymous',
        'anonymous' => true,
        // 'avatar' => 'sample_anonymous_avatar.png'
    ],
];
